### PR TITLE
[cuBLAS][cuBLASLt] Thread-local cuBLAS(Lt) workspace maps

### PR DESCRIPTION
**Pull request by eqy**
Following #167928 and some other more recent PRs that uncovered more than desirable CPU overhead on the cuBLASLt hot path e.g., https://github.com/pytorch/pytorch/pull/179022 https://github.com/pytorch/pytorch/pull/174594 I was thinking about whether we really needed to have cuBLAS(Lt) handles as part of the key to the map for workspaces.

In fact the handle-pool with per-thread handles means that two threads should never use the same cuBLAS(Lt) handle at the same time: https://github.com/pytorch/pytorch/blob/1cb048d50a504298e8f27f950eaa014d91b9adbb/aten/src/ATen/cuda/CublasHandlePool.cpp#L388

With this fact in mind, if we make the workspace map `thread_local`, we can feed two birds with one scone and remove the mutex around the workspace map to potentially save the overhead of (1) needing to get the handle again at all in Lt and (2) acquiring the mutex at all. I don't think this adds any additional workspaces over the previous implementation as separate threads would have had their own workspaces on each device as they had different handles as well, and the overhead per thread of the map is probably on the order of hundreds of bytes. We do need to add the device index as a key to the map as it was previously implicitly tracked by the handle.

authored with claude, the part I'm most unsure about are the changes to the C++ test (should we keep the test?)

Benchmark results soon but I believe this is saving single-digit % E2E on small-medium Lt matmuls on Grace systems.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @azahed98

### CODE REVIEW

File: aten/src/ATen/cuda/CublasHandlePool.cpp [hunk 1]
```diff
@@ -273,95 +256,61 @@ at::DataPtr getNewCUDABlasLtWorkspace() {
...
-
-  // Insert with lock (double-check in case another thread inserted while we
-  // were allocating)
-  {
-    std::unique_lock<std::shared_mutex> lock(workspace.mutex);
-    auto workspace_it = workspace.map.try_emplace(key, std::move(new_workspace)).first;
-    TORCH_CUDABLAS_CHECK(
-        cublasSetWorkspace(handle, workspace_it->second.get(), workspace_size));
-  }
+  auto [it, _] = workspace.emplace(key, getNewWorkspace());
```
Comment (Skylion007) [inline aten/src/ATen/cuda/CublasHandlePool.cpp]: Nit, try_emplace? You can also provide emplace_it as the iterator_hint for even faster insertion via the try_emplace overload or via the emplace_hint method.

**Comment by eqy** [comment]
checked in the temporary benchmarking script for results below

# cuBLAS/cuBLASLt Thread-Local Workspace Benchmark Results

**Platform:** GB200, sm_100 (compute capability 10.0)
**Branches:** `main` (mutex-guarded global workspace map) vs `threadlocalworkspace` (thread_local workspace map)
**Methodology:** Each configuration spawns fresh threads that acquire a cuBLAS handle+workspace or cuBLASLt workspace `calls_per_thread` times, then exit. 3 runs averaged. CUDA runtime TLS is primed on each thread to match real-world usage.

## Table 1: Sequential cuBLAS (`getCurrentCUDABlasHandle`, 1 thread at a time)

| calls/thread | threads | main per-thread (ms) | TLW per-thread (ms) | main per-iter (us) | TLW per-iter (us) | delta |
|---|---|---|---|---|---|---|
| 1 | 20000 | 0.070 | 0.071 | 69.6 | 71.5 | **+2.8%** |
| 10 | 10000 | 0.076 | 0.076 | 7.59 | 7.59 | **0%** |
| 100 | 5000 | 0.109 | 0.108 | 1.09 | 1.07 | **-1.3%** |
| 1000 | 2000 | 0.428 | 0.414 | 0.428 | 0.414 | **-3.2%** |

## Table 2: Sequential cuBLASLt (`getCUDABlasLtWorkspace`, 1 thread at a time)

| calls/thread | threads | main per-thread (ms) | TLW per-thread (ms) | main per-iter (us) | TLW per-iter (us) | delta |
|---|---|---|---|---|---|---|
| 1 | 20000 | 0.068 | 0.068 | 68.2 | 68.4 | **+0.3%** |
| 10 | 10000 | 0.070 | 0.070 | 6.97 | 6.97 | **0%** |
| 100 | 5000 | 0.086 | 0.078 | 0.86 | 0.78 | **-9.5%** |
| 1000 | 2000 | 0.204 | 0.168 | 0.204 | 0.168 | **-17.5%** |

## Table 3: Concurrent cuBLAS (`getCurrentCUDABlasHandle`, 2 threads at a time)

| calls/thread | threads | main per-thread (ms) | TLW per-thread (ms) | main per-iter (us) | TLW per-iter (us) | delta |
|---|---|---|---|---|---|---|
| 1 | 20000 | 0.062 | 0.056 | 61.7 | 55.7 | **-9.8%** |
| 10 | 10000 | 0.059 | 0.058 | 5.93 | 5.76 | **-2.8%** |
| 100 | 5000 | 0.078 | 0.077 | 0.78 | 0.77 | **-1.7%** |
| 1000 | 2000 | 0.579 | 0.530 | 0.579 | 0.530 | **-8.5%** |

## Table 4: Concurrent cuBLASLt (`getCUDABlasLtWorkspace`, 2 threads at a time)

| calls/thread | threads | main per-thread (ms) | TLW per-thread (ms) | main per-iter (us) | TLW per-iter (us) | delta |
|---|---|---|---|---|---|---|
| 1 | 20000 | 0.063 | 0.058 | 63.1 | 58.5 | **-7.2%** |
| 10 | 10000 | 0.061 | 0.060 | 6.15 | 5.96 | **-3.0%** |
| 100 | 5000 | 0.061 | 0.061 | 0.61 | 0.61 | **+0.2%** |
| 1000 | 2000 | 0.309 | 0.222 | 0.309 | 0.222 | **-28.3%** |

## Summary

- **Sequential cold path** (1 call/thread): the optimization is within noise (~0-3%) -- thread creation dominates.
- **Sequential warm path** (1000 calls/thread): cuBLASLt shows a **17.5% improvement** from eliminating `shared_mutex`. cuBLAS shows a **3.2% improvement**.
- **Concurrent** (2 threads at a time): the optimization wins across the board. The cuBLASLt warm path at 1000 calls/thread shows a **28.3% improvement** -- the `shared_mutex` contention from `main` is completely eliminated.
- Per-iteration cost drops dramatically as calls/thread increases (e.g., 69us -> 0.17us for cuBLASLt), confirming thread creation overhead is amortized across iterations.

File: aten/src/ATen/test/cuda_cublas_workspace_bench.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
+#include <gtest/gtest.h>
```
Comment (eqy) [inline aten/src/ATen/test/cuda_cublas_workspace_bench.cpp:1]: vibecoded benchmark, here temporarily so folks an inspect

File: aten/src/ATen/cuda/CUDAContextLight.h [hunk 3]
```diff
@@ -90,14 +90,15 @@ TORCH_CUDA_CPP_API cublasLtHandle_t getCurrentCUDABlasLtHandle();
...
-
-TORCH_CUDA_CPP_API WorkspaceMapWithMutex& cublas_handle_stream_to_workspace();
-TORCH_CUDA_CPP_API WorkspaceMapWithMutex& cublaslt_handle_stream_to_workspace();
-TORCH_CUDA_CPP_API size_t getChosenWorkspaceSize();
+
+// Thread-local workspace map keyed by (device, stream).
+// No mutex is needed because cuBLAS handles are unique per thread
+// (guaranteed by DeviceThreadHandlePool), so each thread's workspace
+// map is only ever accessed by that thread.
+using WorkspaceMap = std::map<std::pair<int, void*>, at::DataPtr>;
```
Comment (ngimel) [inline aten/src/ATen/cuda/CUDAContextLight.h]: should we make this unordered_map?

File: aten/src/ATen/cuda/CublasHandlePool.cpp [hunk 4]
```diff
@@ -273,95 +256,61 @@ at::DataPtr getNewCUDABlasLtWorkspace() {
...
+    auto [it, _] = workspace.emplace(key, getNewWorkspace());
+    return it->second.mutable_get();
   }
 #endif
-  cublasLtHandle_t handle = getCurrentCUDABlasLtHandle();
+  c10::DeviceIndex device = c10::cuda::current_device();
   auto stream = c10::cuda::getCurrentCUDAStream();
   cudaStream_t _stream = stream;
-  auto key = std::make_tuple(static_cast<void *>(handle), static_cast<void *>(_stream));
+  auto key = std::make_pair(static_cast<int>(device), static_cast<void *>(_stream));
```
Comment (ngimel) [inline aten/src/ATen/cuda/CublasHandlePool.cpp:338]: nit: seems you are creating key unconditionally, regardless of rocm/unified settings so pull this code up?

File: aten/src/ATen/cuda/CublasHandlePool.cpp [hunk 5]
```diff
@@ -114,45 +113,29 @@ using CuBlasPoolType = DeviceThreadHandlePool<cublasHandle_t, createCublasHandle
...
-    std::unique_lock<std::shared_mutex> lock(workspace.mutex);
-    workspace.map.clear();
-  }
-  {
-    auto& workspace = cublaslt_handle_stream_to_workspace();
-    std::unique_lock<std::shared_mutex> lock(workspace.mutex);
-    workspace.map.clear();
-  }
+  cublas_stream_to_workspace().clear();
+  cublaslt_stream_to_workspace().clear();
```
Comment (nikitaved) [inline aten/src/ATen/cuda/CublasHandlePool.cpp:135]: Do we really need two maps?

File: aten/src/ATen/cuda/CublasHandlePool.cpp [hunk 6]
```diff
@@ -273,95 +256,61 @@ at::DataPtr getNewCUDABlasLtWorkspace() {
...
-      }
-    }
-    // First use for this handle+stream pair — allocate and insert directly.
-    // No need to call cublasSetWorkspace; Lt passes workspace explicitly.
-    auto new_workspace = getNewWorkspace();
-    {
-      std::unique_lock<std::shared_mutex> lock(workspace.mutex);
-      auto workspace_it = workspace.map.try_emplace(key, std::move(new_workspace)).first;
+    auto key = std::make_pair(static_cast<int>(device), static_cast<void *>(_stream));
+    auto& workspace = at::cuda::cublas_stream_to_workspace();
```
Comment (nikitaved) [inline aten/src/ATen/cuda/CublasHandlePool.cpp]: nit: `workspace -> workspace_map`?

File: aten/src/ATen/cuda/CUDAContextLight.h [hunk 7]
```diff
@@ -95,12 +97,20 @@ TORCH_CUDA_CPP_API void clearCublasWorkspacesForStream(cudaStream_t stream);
 // No mutex is needed because cuBLAS handles are unique per thread
 // (guaranteed by DeviceThreadHandlePool), so each thread's workspace
 // map is only ever accessed by that thread.
-using WorkspaceMap = std::unordered_map<std::pair<int, void*>, at::DataPtr>;
+struct WorkspaceMapHash {
+  size_t operator()(const std::pair<int, void*>& p) const {
+    return std::hash<uintptr_t>{}(
+        reinterpret_cast<uintptr_t>(p.second) ^
+        (static_cast<uintptr_t>(p.first) << 48));
```
Comment (nikitaved) [inline aten/src/ATen/cuda/CUDAContextLight.h]: Is it safe, i.e. any specific reason for 48?

Comment (eqy) [reply aten/src/ATen/cuda/CUDAContextLight.h]: that's a good point, I think the vibecoding here is sketchy because 48-bit wide virtual addresses are a mostly x86 assumption that doesn't hold on SBSA

File: aten/src/ATen/cuda/CUDAContextLight.h [hunk 8]
```diff
@@ -90,16 +92,23 @@ TORCH_CUDA_CPP_API cublasLtHandle_t getCurrentCUDABlasLtHandle();
...
+// Thread-local workspace map keyed by (device, stream).
+// No mutex is needed because cuBLAS handles are unique per thread
+// (guaranteed by DeviceThreadHandlePool), so each thread's workspace
+// map is only ever accessed by that thread.
+struct WorkspaceMapHash {
+  size_t operator()(const std::pair<int, void*>& p) const {
+    return std::hash<int>{}(p.first) ^ std::hash<void*>{}(p.second);
+  }
 };
+using WorkspaceMap = std::unordered_map<std::pair<int, void*>, std::pair<at::DataPtr, size_t>, WorkspaceMapHash>;
```
Comment (ngimel) [inline aten/src/ATen/cuda/CUDAContextLight.h]: why do you need custom hash instead of c10::hash?

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/cuda/CublasHandlePool.cpp (hunk 5 above)
Comment (eqy) [reply aten/src/ATen/cuda/CublasHandlePool.cpp:135]: I think for now we kept the existing structure due to ROCM's diverge lt vs. non-lt handles but this is something we could definitely revisit in the future

**Comment by eqy** [comment]
@ngimel @soulitzer To fix the FSDP issue above caused by autograd threads not clearing their workspaces I vibecoded a solution that makes the user-facing call also issue a callback to autograd threads. Does this seem okay?
I think the user-contract is "you are responsible for clearing workspaces on threads you manage" while autograd threads are not user-managed.

**Comment by eqy** [comment]
@albanD any concerns with autograd thread use here?

File: torch/csrc/autograd/engine.cpp [hunk 9]
```diff
@@ -1525,6 +1540,48 @@ void Engine::queue_callback(std::function<void()> callback) {
   current_graph_task->final_callbacks_.emplace_back(std::move(callback));
 }
 
+// We use this to do housekeeping on-demand e.g., when the user requests
+// cuBLAS workspaces to be cleared
+void Engine::execute_callback_on_device_threads(
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp]: This is a cool trick that claude came up with. But looks very very suspicious tbh.
I'm very surprised you need this at all?

Comment (eqy) [reply torch/csrc/autograd/engine.cpp]: Not needed as in there should be another mechanism to issue a callback or that we shouldn't need to clear workspaces? For the first one I can try to recheck. Second one is basically true, as it's mostly a housekeeping mechanism for tests.

this gadget was written by codex btw

**Comment by ngimel** [comment]
Ok hear me out, `clearCublasWorkspace` as a custom autograd function! Do we need it to be available from c++, or just python will do?

File: torch/csrc/autograd/engine.cpp (hunk 9 above)
Comment (eqy) [reply torch/csrc/autograd/engine.cpp]: OK now it has been moved to use `execute_with_graph_task`

**Comment by eqy** [comment]
@ngimel I guess python only for now is OK? what would that look like e.g.,
```
fake_loss = torch.cuda._clear_cublas_workspaces_after_backward(fake_loss)
```
?

**Comment by eqy** [comment]
Hmm, on second thought while I think autograd function is elegant for places that explicitly have a backward visible such as eager I'm not sure there's a clean way to wire it into stuff that might use the inductor context manager e.g.,
https://github.com/pytorch/pytorch/blob/3920cb139498af9bffbc28c51cfe64ede2c9ebec/torch/_inductor/cudagraph_trees.py#L182

**Comment by ngimel** [comment]
what's wrong with putting it on exit in that context manager? Also it's not really an issue for all inductor-related things because they disable autograd multithreading (as everyone should)

**Comment by eqy** [comment]
maybe I overthought needing access to the loss in the context manager when any dummy loss will do

File: docs/source/notes/cuda.rst [hunk 10]
```diff
@@ -1063,7 +1063,7 @@ cuBLAS workspaces
 For each combination of cuBLAS handle and CUDA stream, a cuBLAS workspace will be allocated
 if that handle and stream combination executes a cuBLAS kernel that requires a workspace.
 In order to avoid repeatedly allocating workspaces, these workspaces are not deallocated unless
-``torch._C._cuda_clearCublasWorkspaces()`` is called. The workspace size per allocation can be
+``torch.cuda._clear_cublas_workspaces()`` is called. The workspace size per allocation can be
```
Comment (albanD) [inline docs/source/notes/cuda.rst]: I'm surprised there isn't a bigger update on this doc. Since we are changing how these workspaces work quite significantly here!

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related, and review the changes in this PR

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@eqy any progress on this?

**Comment by eqy** [comment]
@ngimel I'm not aware of any blockers here, it's orthogonal to churn around workspace sizes right

**Comment by eqy** [comment]
I think the feedback around autograd threads already exited seems overly cautious, those workspaces would have been deleted anyway?

**Comment by ngimel** [comment]
Yeah some comments that claude suggested leaving in the code "Consider a brief inline comment explaining why this is needed (compiled autograd would try to trace the synthetic backward)."

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Comment by eqy** [comment]
@claude explain those CI failures, are they related to this PR?

**Review by jansel — APPROVED** [review]

**Comment by eqy** [comment]
@claude explain the CI failures, are they related to this PR?

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "reverted internally" -c ghfirst

superseding https://github.com/pytorch/pytorch/pull/188685

**Comment by BoyuanFeng** [comment]
@eqy Thanks for the help! We have tested and the previous issue is resolved. Could you plz fix the ci? We can land once the ci is green.

**Comment by eqy** [comment]
@BoyuanFeng remaining failure is preexisting, could you re-import please?
